### PR TITLE
Adding Llama FastTokenizer support.

### PR DESCRIPTION
**Pull request by Narsil**
- Requires https://github.com/huggingface/tokenizers/pull/1183 version
- Only support byte_fallback for llama, raise otherwise (safety net).
- Lots of questions are special tokens

How to test:

```python
#! pip install -e https://github.com/huggingface/tokenizers@byte_fallback#egg=tokenizers

from transformers.convert_slow_tokenizer import convert_slow_tokenizer
from transformers import AutoTokenizer
from tokenizers import Tokenizer

tokenizer = AutoTokenizer.from_pretrained("huggingface/llama-7b")

if False:
    new_tokenizer = Tokenizer.from_file("tok.json")
else:
    new_tokenizer = convert_slow_tokenizer(tokenizer)
    new_tokenizer.save("tok.json")

strings = [
    "This is a test",
    "生活的真谛是",
    "生活的真谛是[MASK]。",
    # XXX: This one is problematic because of special tokens
    # "<s> Something something",
]

for string in strings:
    encoded = tokenizer(string)["input_ids"]
    encoded2 = new_tokenizer.encode(string).ids

    assert encoded == encoded2, f"{encoded} != {encoded2}"

    decoded = tokenizer.decode(encoded)
    decoded2 = new_tokenizer.decode(encoded2)

    assert decoded.strip() == decoded2, f"{repr(decoded)} != {repr(decoded2)}"
```

# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by sgugger** [comment]
Thanks for the ping. We'll need the actual fast tokenizer file to merge this though :sweat_smile:

**Comment by Narsil** [comment]
More troublesome than anticipated.

When encoding `" Hello"` from a pure BPE perspectivve, `tokenizers` does `[259, 10994]`  (`"  "` + `Hello`)
whereas spm does `[29871, 15043]` (`" "` +  `" Hello"`) which from a pure ids & merges perspectives seems worse.

I though of fixing that using a `pre_tokenizer` that splits words onto their own.

However on encoding `"  ird"` this time `spm` DOES do `[259, 1823]`.
Seems this is where the score comes into play.

File: src/transformers/convert_slow_tokenizer.py
Comment (ArthurZucker) [inline src/transformers/convert_slow_tokenizer.py:1114]: Nice!

File: tests/models/llama/test_tokenization_llama.py
Comment (ArthurZucker) [inline tests/models/llama/test_tokenization_llama.py:110]: > - Add a new extra pre_tokenizer which doesn't pretokenize but does this job.
Since this was added we don't need that comment anymore no

File: tests/models/llama/test_tokenization_llama.py
Comment (ArthurZucker) [inline tests/models/llama/test_tokenization_llama.py:127]: Should go away with `cleanup_tokenization_space` here #22341

File: tests/models/llama/test_tokenization_llama.py
Comment (ArthurZucker) [inline tests/models/llama/test_tokenization_llama.py:134]: This will be fixed in the slow tokenizer. 2 things to do:
- open a PR to fix the conversion script 
- open PR on the hub to fix the content of the tokenizer

File: tests/models/llama/test_tokenization_llama.py
Comment (ArthurZucker) [reply tests/models/llama/test_tokenization_llama.py:127]: (flagging to take care of this test if this is merged first)

**Review by ArthurZucker — APPROVED** [review]
Nicely done! 😄 I have to take care of a few things on the slow side and should be done

**Comment by OlivierDehaene** [comment]
What is the status of this PR?

File: tests/models/llama/test_tokenization_llama.py
Comment (ArthurZucker) [inline tests/models/llama/test_tokenization_llama.py:423]: they are no longer different, the merge with the main file was a bit wrong

File: tests/models/llama/test_tokenization_llama.py
Comment (ArthurZucker) [inline tests/models/llama/test_tokenization_llama.py:383]: same here, the full comment should be removed as the solution was implemented

Comment (Narsil) [reply tests/models/llama/test_tokenization_llama.py]: Done

File: tests/models/llama/test_tokenization_llama.py
Comment (Narsil) [reply tests/models/llama/test_tokenization_llama.py]: Done.

File: docs/source/en/model_doc/llama.mdx
Comment (ArthurZucker) [inline docs/source/en/model_doc/llama.mdx:64]:
```suggestion
[[autodoc]] LlamaTokenizerFast
```

File: src/transformers/models/llama/tokenization_llama_fast.py
Comment (ArthurZucker) [inline src/transformers/models/llama/tokenization_llama_fast.py:12]:
```suggestion
```

File: setup.py
Comment (sgugger) [inline setup.py:179]: Will need to be change to a minimum pin.

File: src/transformers/convert_slow_tokenizer.py
Comment (sgugger) [inline src/transformers/convert_slow_tokenizer.py:58]: This needs to be in its own PR with a flag for breaking change.

File: src/transformers/models/llama/tokenization_llama_fast.py
Comment (sgugger) [inline src/transformers/models/llama/tokenization_llama_fast.py:15]: Needs copyright doc etc.

File: tests/models/llama/test_tokenization_llama.py
Comment (sgugger) [inline tests/models/llama/test_tokenization_llama.py:321]: Maybe we need a smaller tokenizer then?

File: src/transformers/convert_slow_tokenizer.py
Comment (Narsil) [reply src/transformers/convert_slow_tokenizer.py:58]: It's not breaking anymore.

Comment (sgugger) [reply src/transformers/convert_slow_tokenizer.py:58]: It still has a very strong potential to be breaking as it touches code functionality, and it will be easier to isolate it in a git bisect if it goes in its own PR. So I insist.
You can just reopen the PR you closed and amend it with those changes.

Comment (Narsil) [reply src/transformers/convert_slow_tokenizer.py:58]: Fair enough.

File: src/transformers/models/llama/tokenization_llama_fast.py
Comment (ArthurZucker) [inline src/transformers/models/llama/tokenization_llama_fast.py]:
```suggestion

require_version("tokenizers>=0.13.3")

```

File: src/transformers/convert_slow_tokenizer.py
Comment (Narsil) [reply src/transformers/convert_slow_tokenizer.py:58]: https://github.com/huggingface/transformers/pull/22582/files

File: src/transformers/models/llama/tokenization_llama_fast.py
Comment (sgugger) [inline src/transformers/models/llama/tokenization_llama_fast.py:1]: Still missing copyright here.

File: src/transformers/models/llama/tokenization_llama_fast.py
Comment (sgugger) [inline src/transformers/models/llama/tokenization_llama_fast.py:10]: Doc could be expanded here

**Review by sgugger — COMMENTED** [review]
Thanks for isolating the change in the conversion in #22582, this PR will need to be rebased after it's merged.

Still one comment on building a smaller tokenizer for the tests if possible and fletching out the fast tokenizer module.

**Comment by Narsil** [comment]
For the doc builder, we're going to need an update on the docker image so that it pulls 0.13.3 to generate the doc.

File: src/transformers/models/llama/tokenization_llama_fast.py
Comment (Narsil) [reply src/transformers/models/llama/tokenization_llama_fast.py:10]: Expanded.

File: src/transformers/models/llama/tokenization_llama_fast.py
Comment (Narsil) [reply src/transformers/models/llama/tokenization_llama_fast.py:17]: I did add the special tokens.

I have no idea what the args are supposed to be.`PreTrainedTokenizerFast` is also using `*args`.

**Review by sgugger — APPROVED** [review]
Good to go once all tests pass. Thanks!

**Comment by stefan-it** [comment]
Hi @Narsil ,

the change from `warning.warn` to `raise RuntimeError` change in `src/transformers/convert_slow_tokenizer.py` breaks a lot of things: I wanted to fine-tune a mT5 model and it is now no longer possible (I'm using the PyTorch example from [documentation](https://github.com/huggingface/transformers/tree/main/examples/pytorch/question-answering#fine-tuning-t5-on-squad20).)

How is it possible to rubustify it -> also DeBERTa v3 has byte fallback vocab (but I didn't test it yet) :thinking:

**Comment by Narsil** [comment]
> Hi @Narsil ,
> 
> the `warning.warn` to `raise RuntimeError` change in `src/transformers/convert_slow_tokenizer.py` breaks a lot of things: I wanted to fine-tune a mT5 model and it is now no longer possible (I'm using the PyTorch example from 

 How is it possible to rubustify it -> also DeBERTa v3 has byte fallback vocab (but I didn't test it yet) thinking


First of all we could revert by all means, but since now `tokenizers` has `ByteFallback` we could make it 1-1 for those, that was the idea behind upping to an error.

It's a relatively sizeable issue if there are models deployed out there which have inconsistent behavior regarding this though (slow using byte fallback, fast not using it). I'm not sure why it was a warning in the first place.


>  DeBERTa v3 

Let's have a look too.

As a user, what's your opinion here, should we just fix the various conversion scripts, or would you rather keep the warning with the previous pitfalls ?

**Comment by Narsil** [comment]
Both are using Unigram with ByteFallback which isn't supported yet.

**Comment by fxmarty** [comment]
@Narsil After this commit `AutoTokenizer.from_pretrained` is extremely slow, spending time in `convert_slow_tokenizer.py` at every call. Is it expected?

**Comment by Narsil** [comment]
Which repo are you using? We need to create the fast files on the repo.


Converting from slow is super slow and there's nothing to be done about it (tokenizers needs to recreate a structure by doing O(n2) search over the vocab because spm does not store this information.

**Comment by Narsil** [comment]
@ArthurZucker

**Comment by fxmarty** [comment]
I see thanks!

**Pull request merged**
